056-zk-kraft-migration.md

```diff
@@ -0,0 +1,258 @@
+
+#### Phase 5: Removing ZooKeeper
+
+At this point the ZooKeeper ensemble is out of the picture, still running but not used anymore.
+In the `FullKRaft` state, the operator expects the user to finalize the migration by removing ZooKeeper.
+The user removes the `spec.zookeeper` on the `Kafka` custom resource so that the operator can delete the ZooKeeper nodes.
+The FSM moves back to a `NoMigration` state (or null) and the operator removes the `strimzi.io/kraft-migration-state` annotation from the `Kafka` custom resource.
+
+### Metrics
+
+During ZooKeeper to KRaft migration, there are some metrics that could be useful to check the status in each of the phases.
+More details in the [“Metrics”](https://cwiki.apache.org/confluence/display/KAFKA/KIP-866+ZooKeeper+to+KRaft+Migration#KIP866ZooKeepertoKRaftMigration-Metrics) section of KIP-866.
+
+The most interesting and useful ones are:
+
+| MBean                                                             | Description                                                                                                                        |
+|-------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------|
+| kafka.controller:type=KafkaController,name=MetadataType           | An enumeration of: ZooKeeper (1), KRaft (2), or Dual (3). The active controller reports this.                                      |
+| kafka.controller:type=KafkaController,name=MigratingZkBrokerCount | A count of ZK brokers that are registered with KRaft and ready for migration. This will only be reported by the active controller. |
+| kafka.controller:type=KafkaController,name=ZkMigrationState       | An enumeration of the possible migration states the cluster can be in. This is only reported by the active controller.             |
+
+They could be used the following way:
+
+* `MetadataType`: to understand if the active controller is not in the "dual write" mode anymore (keeping metadata in both ZooKeeper and KRaft) and also out of ZooKeeper, keeping metadata only as KRaft.
+* `MigratingZkBrokerCount`: to understand when the migration is starting because after creating the KRaft controllers and forming the quorum, they will "wait for brokers" (beginning of Phase 2). It happens when brokers are switched to the migration mode. Having all brokers registered means migration was started/is going to start.
+* `ZkMigrationState`: to get the actual status of the migration. More details [here](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/metadata/migration/ZkMigrationState.java).
+
+The Kafka Agent could expose a `/kraft-migration` endpoint providing some of/all the metrics above (in JSON format) which are useful during the automated migration process, in order to allow the operator to understand when a phase is completed and the user can drive to the next one.
```

Comment:
Could? Will the operator not use it?

Reply:
I will change it as "confirmation" to use it :-)

Reply:
Are you still going to change this @ppatierno, because I had the same comment as Jakub.